keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,374 @@
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+
+## Motivation
+
+JobSet's current FailurePolicy API only allows specifying the max number of JobSet restarts to attempt before marking it as failed.
+
+Users have requested that we should also allow them to configure the failure conditions under which the JobSet will be restarted, or when the JobSet should be failed immediately without going through the allowed number of restarts. 
+
+This will allow users to make more efficient use of their computing resources, by avoiding `maxRestarts` number of unnecessary 
+JobSet restarts in the case of a non-retriable failure (e.g., an application code bug), while still restarting in the event of
+retriable failures (e.g., a host maintenace event).
+
+This is especially important for large scale distributed training workloads on expensive hardware accelerators. Recreating the 
+entire JobSet may take ~1 minute or so (per latest scale tests on several node pools of ~3k nodes each), and if `maxRestarts`
+is set to a very high number to allow for disruptions (which occur often at a large scale like this), the workload could
+waste roughly `maxRestarts` number of minutes repeatedly recreating despite the fact that it is doomed to fail due to a non-retriable
+error like an application code bug. This is very expensive and wasteful on expensive, large scale clusters.
+
+We need to define a more configurable Failure Policy API which will allow users to define their own restart and failure semantics, 
+based on the type of child job failure occurring.
+
+### Goals
+
+- Enhance the Failure Policy API with options allowing the user to control the restart and failure semantics of the JobSet.
+
+### Non-Goals
+
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, in order to use my computing resources as efficiently as possible, I want to 
+configure my JobSet to restart in the event of a child job failure due to a retriable error
+like host maintance, but to fail the JobSet immediately without any unnecessary restart 
+attempts in the event of an non-retriable application code error. 
+
+**Example Failure Policy Configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: configurable-failure-policy-1
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: cloud.google.com/gke-nodepool # 1:1 job replica to node pool assignment
```

Comment:
I would suggest leaving out this if its not necessary. Project should be cloud neutral.

Reply:
Updated